File: observer.cpp
/* https://p.ip.fi/iYhp
use case: whenever there is a data change happening and many multiple tasks should be carried out.

in other way - when a publisher publishes an event, all subscribers for a topic consume it

UML diagram - https://imgur.com/a/0eoz9aT


checkout this YT short too - https://youtube.com/shorts/K1HHbKJvcv8?si=3S2rBVvjbFxzT1dK (no need of below interface)

to extend this further as a followup, there can be push or pull type of Subject(observerable)


Here Group class is Subject (observable), and User is observer

*/

#include <iostream>
#include <set>
#include <string>

using namespace std;

class ISubscriber {
  protected:
    string name;
  public:
    virtual void notify(string message) = 0;
};

class User: public ISubscriber {
  public:
    User(string name) {
      this->name = name;
    }
    void notify(string message) {
      cout << name << " is notified about: " << message << "\n";
    }
};
// there can be different implementation of ISubscriber to subscribe other than use User class

class Group {
  private:
    set<ISubscriber*> subscribers;
  public:
    void addSubscriber(ISubscriber &sub) {
      subscribers.insert(&sub);
    }

    void removeSubscriber (ISubscriber &sub) {
      subscribers.erase(&sub);
    }

    void notify(string message) {
      for (ISubscriber* sub: subscribers) {
        sub->notify(message);
      }
    }
};

int main () {

  Group g;

  User user1("sid");
  User user2("mom");
  User user3("sonu");

  g.addSubscriber(user1);
  g.addSubscriber(user2);
  g.addSubscriber(user3);

  g.notify("message1");

  g.removeSubscriber(user3);

  g.notify("message2");

  return 0;

}

File: design_patterns/typescript/observer.ts
/* https://p.ip.fi/iYhp
use case: whenever there is a data change happening and many multiple tasks should be carried out.

in other way - when a publisher publishes an event, all subscribers for a topic consume it

UML diagram - https://imgur.com/a/0eoz9aT

checkout this YT short too - https://youtube.com/shorts/K1HHbKJvcv8?si=3S2rBVvjbFxzT1dK (no need of below interface)

to extend this further as a followup, there can be push or pull type of Subject(observerable)

Here Group class is Subject (observable), and Person is observer
*/

// Observer interface
interface ISubscriber {
    notify(message: string): void;
    getName(): string;
}
interface Person {
    getName(): string;
}

//Key principle: TypeScript only cares if the object has at least what's required by the interface. Extra properties/methods are fine
// Concrete Observer
class Employee implements Person, ISubscriber {
    private name: string;

    constructor(name: string) {
        this.name = name;
    }

    notify(message: string): void {
        console.log(`${this.name} is notified about: ${message}`);
    }

    getName(): string {
        return this.name;
    }
}

// Subject (Observable)
class Group {
    private subscribers: Set<ISubscriber> = new Set();

    addSubscriber(subscriber: ISubscriber): void {
        this.subscribers.add(subscriber);
        console.log(`${subscriber.getName()} subscribed to the group`);
    }

    removeSubscriber(subscriber: ISubscriber): void {
        if (this.subscribers.delete(subscriber)) {
            console.log(`${subscriber.getName()} unsubscribed from the group`);
        }
    }

    notify(message: string): void {
        console.log(`\nBroadcasting: "${message}"`);
        this.subscribers.forEach(subscriber => {
            subscriber.notify(message);
        });
    }
}

// Main function
function main(): void {
    const group = new Group();

    // Create users
    const user1 = new Employee("sid");
    const user2 = new Employee("mom");
    const user3 = new Employee("sonu");

    // Subscribe users to group
    group.addSubscriber(user1);
    group.addSubscriber(user2);
    group.addSubscriber(user3);

    // Send first notification
    group.notify("message1");

    // Remove user3
    group.removeSubscriber(user3);

    // Send second notification
    group.notify("message2");
}

// Execute main
main();